Add derivation paths to coin selection change outputs

jonathanknowles opened the pull request:
# Issue Number

#2247 
  
# Overview

This PR adjusts the API [`selectCoins`](https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/selectCoins) operation so that:

- _ordinary_ outputs and _change_ outputs are held in separate fields in the returned selection:

    | Field | Description |
    | -- | -- |
    | **`outputs`** | ordinary payments to external wallets |
    | **`change`** | change to be returned to the wallet |
- each change output now has a `derivation_path` field containing the full derivation path for the change output address.

  The format of this field is identical to the format of the existing `derivation_path` field of each input.

<details><summary>Click here to see updated API description</summary>

![separate-change-ouputs](https://user-images.githubusercontent.com/206319/96819250-bcf6d200-1455-11eb-88c7-3135265f8cd1.png)

</details>

# Internal Changes

This PR introduces a `ApiCoinSelectionChange` type that includes a derivation path:

```hs
data ApiCoinSelectionChange (n :: NetworkDiscriminant) = ApiCoinSelectionChange
    { address :: !(ApiT Address, Proxy n)
    , amount :: !(Quantity "lovelace" Natural)
    , derivationPath :: NonEmpty (ApiT DerivationIndex)
    } deriving (Eq, Generic, Show)
```

In addition, we modify the `ApiCoinSelection` type so that ordinary outputs (**_without_** derivation paths) and change outputs (**_with_** derivation paths) are held in separate fields:

```patch

data ApiCoinSelection (n :: NetworkDiscriminant) = ApiCoinSelection
    { inputs :: !(NonEmpty (ApiCoinSelectionInput n))
-   , outputs :: !(NonEmpty (AddressAmount (ApiT Address, Proxy n)))
+   , outputs :: !(NonEmpty (ApiCoinSelectionOutput n))
+   , change :: ![ApiCoinSelectionChange n]
    } deriving (Eq, Generic, Show)

```
# Testing

This PR updates the following integration tests:

- `WALLETS_COIN_SELECTION_01`.
- `WALLETS_COIN_SELECTION_02`.
- `SHELLEY_HW_WALLETS/HW_WALLETS_04`.
- `BYRON_HW_WALLETS/HW_WALLETS_04`.

jonathanknowles committed:
Adjust code to comply with team coding standards.

jonathanknowles committed:
Add type `ApiCoinSelectionChange`.

Add type `ApiCoinSelectionChange`, along with JSON instances and
roundtrip tests.

Values of this type will be used to encode change outputs, within
`ApiCoinSelection` objects.

jonathanknowles committed:
Add field `change` to `ApiCoinSelection`.

This field will eventually hold the change outputs for a coin selection,
allowing them to be easily distinguished from ordinary outputs.
(Currently, both change outputs and ordinary outputs are stored in the
`outputs` field.)

For the moment, this field is hardwired to be empty.

jonathanknowles committed:
Add `unsignedChange` field to `UnsignedTx`.

This allows change outputs to be distinguished from ordinary outputs
within an `UnsignedTx`, which makes it easy to construct an
`ApiCoinSelection` where change outputs are distinguished from ordinary
outputs.

jonathanknowles committed:
Populate the `change` field of `ApiCoinSelection`.

Populate the `change` field of `ApiCoinSelection` within
`mkApiCoinSelection`.

jonathanknowles committed:
Split function `assignChangeAddresses`.

Split function `assignChangeAddresses` into:

- outer function: `assignChangeAddressesForSelection`
- inner function: `assignChangeAddresses`

This allows the inner function `assignChangeAddresses` to be called from
other functions.

jonathanknowles committed:
Make `selectCoinsExternal` populate the `change` field of `UnsignedTx`.

jonathanknowles committed:
Introduce type `ApiCoinSelectionOutput`.

Use this type for the `outputs` field of the `ApiCoinSelection` type.

After applying this change, we have three types with a consistent naming
scheme:

- `ApiCoinSelectionInput`
- `ApiCoinSelectionOutput`
- `ApiCounSelectionChange`

jonathanknowles committed:
Use `Void` type in function `signTx` for added safety.

This function is currently only used in contexts where all change outputs
have been assigned with addresses and are included in the set of ordinary
outputs. We use the 'Void' type here to prevent callers from accidentally
passing change values into this function.

jonathanknowles committed:
Update integration test `WALLETS_COIN_SELECTION_01`.

Update integration test `WALLETS_COIN_SELECTION_01` to check that:

- change outputs are listed within the `change` field of an `ApiCoinSelection`.
- change output derivation paths are valid.

jonathanknowles committed:
Update integration test `WALLETS_COIN_SELECTION_02`.

Update integration test `WALLETS_COIN_SELECTION_02` to check that the
set of coin selection outputs is exactly the same as intially provided,
rather than being a subset.

jonathanknowles committed:
Update integration test `SHELLEY_HW_WALLETS/HW_WALLETS_04`.

Update integration test `SHELLEY_HW_WALLETS/HW_WALLETS_04` to make
multiple payments.

In addition, we expect that the resulting selection will have:

* an output for each of the original payments.
* at least one change output.

jonathanknowles committed:
Update integration test `BYRON_HW_WALLETS/HW_WALLETS_04`.

Update integration test `BYRON_HW_WALLETS/HW_WALLETS_04` to make
multiple payments.

In addition, we expect that the resulting selection will have:

* an output for each of the original payments.
* at least one change output.

paweljakubas commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1624,26 +1633,34 @@ signTx
+    -- have been assigned with addresses and are included in the set of ordinary
+    -- outputs. We use the 'Void' type here to prevent callers from accidentally
+    -- passing change values into this function:
+    -> UnsignedTx input output Void
```
for quite a while have not seen Void :-)

jonathanknowles replied:
Originally I wrote this by specifying `change ~ Void` in the list of constraints, and then specifying `UnsignedTx input output change` as the parameter type, as I felt that this would be more readable.

However, GHC views this as a redundant constraint. Of course, it's possible to silence the redundant constraint warning, but that would mean silencing this warning for the whole file. So I eventually switched to just writing `Void` inline, even though it is slightly less readable (IMO).

paweljakubas commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -1575,22 +1579,27 @@ migrateWallet ctx (ApiT wid) migrateData = do
 -- recycled in order of their appearance in the original list.
 assignMigrationAddresses
-    :: [Address]
+    :: forall input output change .
```
I wonder why to introduce type parametrization and use type equality (very concrete like `output ~TxOut`) rather than stick to what we had?

paweljakubas replied:
this is one place, but this technique was adopted in many places in this PR

jonathanknowles replied:
> I wonder why to introduce type parametrization and use type equality (very concrete like output ~TxOut) rather than stick to what we had?

It's just an attempt to improve readability.

Otherwise, we'd have to write things like:
```hs
UnsignedTx (TxIn, TxOut, NonEmpty DerivationIndex) TxOut (TxChange (NonEmpty DerivationIndex))
```
Which gets repetitive very quickly.

Giving short names to these types also means that we can reuse them within the signatures of inner functions. See `mkApiCoinSelection` for an example of this:

```hs
mkApiCoinSelection
    :: forall n input output change.
        ( input ~ (TxIn, TxOut, NonEmpty DerivationIndex)
        , output ~ TxOut
        , change ~ TxChange (NonEmpty DerivationIndex)
        )
    => UnsignedTx input output change
    -> ApiCoinSelection n
mkApiCoinSelection (UnsignedTx inputs outputs change) =
    ApiCoinSelection
        (mkApiCoinSelectionInput <$> inputs)
        (mkApiCoinSelectionOutput <$> outputs)
        (mkApiCoinSelectionChange <$> change)
  where
    mkApiCoinSelectionInput :: input -> ApiCoinSelectionInput n
    mkApiCoinSelectionInput (TxIn txid index, TxOut addr (Coin c), path) =
        ...
    mkApiCoinSelectionOutput :: output -> ApiCoinSelectionOutput n
    mkApiCoinSelectionOutput (TxOut addr (Coin c)) =
        ...
    mkApiCoinSelectionChange :: change -> ApiCoinSelectionChange n
    mkApiCoinSelectionChange txChange =
        ...
```

paweljakubas reviewed (approved):
so far lgtm.having a way to derive change and inputs begs for some advanced testing sometime in future: we have a wallet with mnemonic1 with one utxo, and send some money to another one. We can now have a handle to output and change. We re-derive the addresses using derivation paths in cardano-addresses. And use them to create another transaction blob in cardano-transactions. Then send it using external api endpoint and see that it works in integration testing . Maybe @piotr-iohk would be interested in adding such a test sometime in the future:-)

jonathanknowles committed:
Add function `qualifyAddresses`.

Use this function to factor out common parts of `selectCoinsExternal`.

jonathanknowles committed:
Move creation of unsigned TX into atomically block.

jonathanknowles committed:
Simplify error type synonyms.

jonathanknowles committed:
Use `Void` type in function `assignMigrationAddresses`.

Use `Void` type in function `assignMigrationAddresses` to indicate that
no change is returned.

jonathanknowles committed:
Remove redundant constraints.

jonathanknowles committed:
Use the payment key when generating a payment address.

Before this change we were constructing payment addresses from the stake
key, which remains constant over time.

This resulted in the generation of a constant "payment address" that was
not recognizable as a valid payment address belonging to the wallet,
causing subsequent calls to `isOurs` to fail.

jonathanknowles committed:
Update integration test `STAKE_POOLS_JOIN_UNSIGNED_01`.

We now expect that there are:

* no ordinary outputs.
* at least one change output.

jonathanknowles committed:
Update integration test `STAKE_POOLS_JOIN_UNSIGNED_02`.

We now expect that there are:

* no ordinary outputs.
* at least one change output.

jonathanknowles committed:
Update integration test `STAKE_POOLS_QUIT_UNSIGNED_01`.

We now expect that there are:

* no ordinary outputs.
* at least one change output.
